rust/src/json_serialize.rs

```diff
@@ -0,0 +1,996 @@
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: Serializer {
+        let string = self.clone().to_string().map_err(|err| serde::ser::Error::custom(format!("serialize: {:?}", err)))?;
+
+        serializer.serialize_str(string.as_str())
+    }
+}
+
+impl<'de> Deserialize<'de> for Value {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: Deserializer<'de> {
+        let s = <String as serde::de::Deserialize>::deserialize(deserializer)?;
+        Value::from_string(s).map_err(|err| serde::de::Error::custom(format!("Can't deserialize json string: {}", err)))
+    }
+}
+
+#[derive(Debug, Clone)]
+enum JsonParseError {
+    // general
+    InvalidToken(JsonToken),
+    InvalidParseResult(Vec<JsonToken>),
+
+    // array and object
+    InvalidTokenBeforeArrayOrObjectStart(JsonToken),
+
+    // array
+    NotAllowedInArray(JsonToken),
+    NoArrayStartFound,
+    ArrayCommaError,
+
+    // object
+    NotAllowedInObject(JsonToken),
+    NoObjectStartFound,
+    ObjectStructureError,
+    NoValueForKey(JsonToken),
+    NoKeyForValue(JsonToken),
+
+    // quote
+    InvalidTokenBeforeQuote(JsonToken),
+    // colon
+    InvalidTokenBeforeColon(Option<JsonToken>),
+    // comma
+    InvalidTokenBeforeComma(Option<JsonToken>),
+    // string
+    InvalidTokenBeforeString(JsonToken),
+    InvalidRawString(String),
+}
+
+impl Display for JsonParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.serialize_str(format!("{:?}", self).as_str())
+    }
+}
+
+fn tokenize_string(string: String) -> Vec<JsonToken> {
+    let mut tokens = Vec::<JsonToken>::new();
+    let mut current_string: String = String::new();
+    for char in string.graphemes(true) {
+        let (reset_string, token) = match char {
+            "\"" => {
+                if !current_string.is_empty() && current_string.clone().graphemes(true).last().clone().unwrap() == "\\" {
+                    let graphemes_count = current_string.clone().graphemes(true).count();
+                    current_string = current_string.graphemes(true).take(graphemes_count - 1).collect();
+                    current_string += "\"";
+                    (false, None)
+                } else {
+                    (true, Some(JsonToken::Quote))
+                }
+            }
+            "{" => {
+                (true, Some(JsonToken::ObjectStart))
+            }
+            "}" => {
+                (true, Some(JsonToken::ObjectEnd))
+            }
+            "[" => {
+                (true, Some(JsonToken::ArrayStart))
+            }
+            "]" => {
+                (true, Some(JsonToken::ArrayEnd))
+            }
+            ":" => {
+                (true, Some(JsonToken::Colon))
+            }
+            "," => {
+                (true, Some(JsonToken::Comma))
+            }
+            _ => {
+                let splitted: Vec<&str> = char.split_whitespace().into_iter().collect();
```

Comment:
that part of the code is not beautiful, sorry about that. that was the easiest way to get rid of whitespaces

Reply:
unfortunately no, technically whitespace is a valid symbol in strings

Reply:
Oh yeah of course that makes sense.